main.go

```diff
@@ -12,13 +18,27 @@ type ProcessSettings struct {
 	FlushAfter int
 }
 
-// FirehoseBackend is used for consuming messages from a transport and read/write to storage
+// ReceivedMessage is the message as received by a transport backend.
+type ReceivedMessage struct {
+	Payload          []byte
+	Attributes       map[string]string
+	ProviderMetadata interface{}
+}
+
+// FirehoseBackend is used for read/write to storage
 type FirehoseBackend interface {
+	UploadFile(ctx context.Context, data []byte, uploadBucket string, uploadLocation string) error
```

Comment:
wonder if upload bucket and upload location can be part of backend's configuration?

Reply:
yeah looking at it now the backend functions will be passing in the locations to the readfile and uploadfile funcs so guess it should go there instead

Reply:
Well, i mean the bucket name sounds like it could go into `NewBackend` function, this way you're not passing them every time you want to save a file. The filename probably should be part of this interface since the time windowing is done here.

Reply:
oh so I guess I should have UploadFollowerFile and UploadLeaderFile then? Since they output differently?  Or will they output to same filepath just a different bucket?

Reply:
I guess on the read side only the leader will have to read from gcs since its just reading the files from the followers?  But then we'd also need a read leader file function for when its checking if its the leader right?

Reply:
oh i forgot about leader and follower reading different files. hmm. Yeah I guess this interface makes sense then. Better than bleeding leader/follower abstraction into storage layer.